### PR TITLE
[IOTDB-5091] The multi-tenancy design about iotdb - add throtttle quota

### PR DESCRIPTION
**Pull request by RYH61**
see [https://issues.apache.org/jira/browse/IOTDB-5091?filter=-1](https://issues.apache.org/jira/browse/IOTDB-5091?filter=-1)

image

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/quotas/DataNodeThrottleQuotaManager.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  private static final Logger LOGGER = LoggerFactory.getLogger(DataNodeThrottleQuotaManager.class);
+
+  private ThrottleQuotaLimit throttleQuotaLimit;
+
+  public DataNodeThrottleQuotaManager() {
+    throttleQuotaLimit = new ThrottleQuotaLimit();
+    recover();
+  }
+
+  /** SingleTone */
```
Comment (mychaow) [inline server/src/main/java/org/apache/iotdb/db/quotas/DataNodeThrottleQuotaManager.java:46]: singleton

**Review by mychaow — APPROVED** [review]
LGTM, only some minor problems.

File: node-commons/src/main/java/org/apache/iotdb/commons/conf/CommonDescriptor.java [hunk 2]
```diff
@@ -191,6 +191,9 @@ public void loadCommonProps(Properties properties) {
         Boolean.parseBoolean(
             properties.getProperty("quota_enable", String.valueOf(config.isQuotaEnable()))));
 
+    config.setRateLimiterType(
+        properties.getProperty("rate_limiter_type", config.getRateLimiterType()));
```
Comment (neuyilan) [inline node-commons/src/main/java/org/apache/iotdb/commons/conf/CommonDescriptor.java:195]: This configuration is not exposed in the iotdb-common.properties?

File: server/src/main/java/org/apache/iotdb/db/metadata/mtree/MTreeBelowSGMemoryImpl.java [hunk 3]
```diff
@@ -384,7 +384,7 @@ private IMemMNode checkAndAutoCreateDeviceNode(String deviceName, IMemMNode devi
       if (!DataNodeSpaceQuotaManager.getInstance().checkDeviceLimit(storageGroupMNode.getName())) {
         throw new ExceedQuotaException(
             "The number of devices has reached the upper limit",
-            TSStatusCode.EXCEED_QUOTA_ERROR.getStatusCode());
+            TSStatusCode.SPACE_QUOTA_EXCEEDED.getStatusCode());
       }
```
Comment (neuyilan) [inline server/src/main/java/org/apache/iotdb/db/metadata/mtree/MTreeBelowSGMemoryImpl.java:388]: as we have the `quota_enable` configuration, I think we can check the parameter first when we check the space/throttle quota.
And I see there are two places that contains the `quota_enable` configurtion, the first is in the CommonDescriptor, the second is in the  IoTDBDescriptor, is it really needed in two places？

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/driver/DataDriver.java [hunk 4]
```diff
@@ -39,9 +39,11 @@
 public class DataDriver extends Driver {
 
   private boolean init;
+  private final long estimatedMemorySize;
```
Comment (neuyilan) [inline server/src/main/java/org/apache/iotdb/db/mpp/execution/driver/DataDriver.java:42]: add some comments about the unit of the size, byte,kb or mb?

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/AnalyzeVisitor.java [hunk 5]
```diff
@@ -243,7 +243,7 @@ public Analysis visitQuery(QueryStatement queryStatement, MPPQueryContext contex
       if (queryStatement.isGroupByTag()) {
         schemaTree = schemaFetcher.fetchSchemaWithTags(patternTree);
       } else {
-        schemaTree = schemaFetcher.fetchSchema(patternTree);
+        schemaTree = schemaFetcher.fetchSchema(patternTree, context);
       }
```
Comment (neuyilan) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/AnalyzeVisitor.java:247]: what is the `context` usage? the same as blow

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/schema/ClusterSchemaFetcher.java [hunk 6]
```diff
@@ -54,6 +55,7 @@ public class ClusterSchemaFetcher implements ISchemaFetcher {
   private final Coordinator coordinator = Coordinator.getInstance();
   private final DataNodeSchemaCache schemaCache = DataNodeSchemaCache.getInstance();
   private final ITemplateManager templateManager = ClusterTemplateManager.getInstance();
+  private MPPQueryContext context;
```
Comment (neuyilan) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/schema/ClusterSchemaFetcher.java:58]: ths same

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LocalExecutionPlanner.java [hunk 7]
```diff
@@ -141,6 +141,7 @@ private void checkMemory(Operator root, FragmentInstanceStateMachine stateMachin
             }
           }
         });
+    return estimatedMemorySize;
```
Comment (neuyilan) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/LocalExecutionPlanner.java:144]: If quota_enable is true, I think we should estimate the memory usage.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/AnalyzeVisitor.java (hunk 5 above)
Comment (RYH61) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/AnalyzeVisitor.java:247]: The context contains information about the session

Comment (neuyilan) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/AnalyzeVisitor.java:247]: However, I do not see it's usage.

**Review by neuyilan — APPROVED** [review]
LGTM

**Pull request merged**